server/src/main/java/io/druid/client/selector/ServerSelector.java

```diff
@@ -62,59 +66,88 @@ public void addServerAndUpdateSegment(
-      if (priorityServers == null) {
-        return false;
+      if (server.getServer().getType() == ServerType.REALTIME) {
+        return realtimeServers.computeIfPresent(
+            server.getServer().getPriority(),
+            (p, servers) -> servers.remove(server) ? servers : null
+        ) == null ? false : true;
+      } else if (server.getServer().getType() == ServerType.HISTORICAL) {
+        return historicalServers.computeIfPresent(
+            server.getServer().getPriority(),
+            (p, servers) -> servers.remove(server) ? servers : null
+        ) == null ? false : true;
       }
-      boolean result = priorityServers.remove(server);
-      if (priorityServers.isEmpty()) {
-        servers.remove(priority);
-      }
-      return result;
     }
+    return false;
   }
 
   public boolean isEmpty()
   {
     synchronized (this) {
-      return servers.isEmpty();
+      return historicalServers.isEmpty() && realtimeServers.isEmpty();
     }
   }
 
   public List<DruidServerMetadata> getCandidates(final int numCandidates)
   {
+    List<DruidServerMetadata> candidates;
     synchronized (this) {
       if (numCandidates > 0) {
-        return strategy.pick(servers, segment.get(), numCandidates)
-                       .stream()
-                       .map(server -> server.getServer().getMetadata())
-                       .collect(Collectors.toList());
+        candidates = new ArrayList<>(numCandidates);
+        strategy.pick(historicalServers, segment.get(), numCandidates)
+            .stream()
+            .map(server -> server.getServer().getMetadata())
+            .forEach(candidates::add);
+
+        if (candidates.size() < numCandidates) {
+          strategy.pick(realtimeServers, segment.get(), numCandidates - candidates.size())
+              .stream()
+              .map(server -> server.getServer().getMetadata())
+              .forEach(candidates::add);
+        }
       } else {
+        candidates = new ArrayList<>();
         // return all servers as candidates
-        return servers.values()
-                      .stream()
-                      .flatMap(Collection::stream)
-                      .map(server -> server.getServer().getMetadata())
-                      .collect(Collectors.toList());
+        historicalServers.values()
+            .stream()
+            .flatMap(Collection::stream)
+            .map(server -> server.getServer().getMetadata())
+            .forEach(candidates::add);
+
+        realtimeServers.values()
+            .stream()
+            .flatMap(Collection::stream)
+            .map(server -> server.getServer().getMetadata())
+            .forEach(candidates::add);
       }
+
+      return candidates;
     }
   }
 
   @Override
   public QueryableDruidServer pick()
   {
     synchronized (this) {
-      return strategy.pick(servers, segment.get());
+      if (!historicalServers.isEmpty()) {
+        return strategy.pick(historicalServers, segment.get());
```

Comment:
What will happen if the strategy is not able to pick from historicals (throws exception), but is able to pick from realtime servers? Is not theoretically?

Reply:
Comments on this?

Reply:
Hrmm, I don't see a reason why picking from historical will result in an NPE and not from realtime. I think this logic should work.

Reply:
It could happen if there are no historicals but there are some realtimes

Reply:
@niketh please answer here

Reply:
@leventov If there are no historicals then `historicalServers.isEmpty()` will be true and we will check realtimes right according to the code. 

Did I misunderstand your question?